File: internal/mcp/biology/sequence.go
package biology

import (
	"fmt"
	"strings"
	"unicode"
)

func maxAnalyzeLength() int {
	return biologySettings().MaxAnalyzeLengthOrDefault()
}

func maxFoldLength() int {
	return biologySettings().MaxFoldLengthOrDefault()
}

// normalizeSequence strips FASTA headers/whitespace and uppercases.
func normalizeSequence(raw string) string {
	var b strings.Builder
	for _, line := range strings.Split(raw, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, ">") || strings.HasPrefix(line, ";") {
			continue
		}
		b.WriteString(line)
	}
	return strings.ToUpper(strings.TrimSpace(b.String()))
}

type seqKind int

const (
	seqUnknown seqKind = iota
	seqDNA
	seqRNA
	seqProtein
)

func classifySequence(seq string) seqKind {
	if seq == "" {
		return seqUnknown
	}
	dna, rna, protein, invalid := 0, 0, 0, 0
	for _, r := range seq {
		switch r {
		case 'A', 'C', 'G', 'T':
			dna++
		case 'U':
			rna++
		case 'N':
			// ambiguous — count toward all
			dna++
			rna++
		case '*':
			protein++
		default:
			if strings.ContainsRune("ACDEFGHIKLMNPQRSTVWY", r) {
				protein++
			} else {
				invalid++
			}
		}
	}
	if invalid > 0 && invalid > len(seq)/10 {
		return seqUnknown
	}
	if protein > 0 && protein >= dna && protein >= rna {
		return seqProtein
	}
	if rna > dna && rna > 0 {
		return seqRNA
	}
	if dna > 0 {
		return seqDNA
	}
	return seqUnknown
}

func reverseComplementDNA(seq string) string {
	complement := map[byte]byte{
		'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N',
	}
	out := make([]byte, len(seq))
	for i := len(seq) - 1; i >= 0; i-- {
		c := seq[i]
		if v, ok := complement[c]; ok {
			out[len(seq)-1-i] = v
		} else {
			out[len(seq)-1-i] = c
		}
	}
	return string(out)
}

func analyzeSequenceText(raw string) (string, error) {
	seq := normalizeSequence(raw)
	if seq == "" {
		return "", fmt.Errorf("no sequence found (paste FASTA or raw sequence)")
	}
	maxLen := maxAnalyzeLength()
	truncated := false
	if len(seq) > maxLen {
		seq = seq[:maxLen]
		truncated = true
	}

	kind := classifySequence(seq)
	var kindStr string
	switch kind {
	case seqDNA:
		kindStr = "DNA"
	case seqRNA:
		kindStr = "RNA"
	case seqProtein:
		kindStr = "protein"
	default:
		kindStr = "unknown/ambiguous"
	}

	var invalid []string
	for i, r := range seq {
		if kind == seqProtein {
			if !unicode.IsLetter(r) && r != '*' {
				invalid = append(invalid, fmt.Sprintf("pos %d: %q", i+1, r))
			}
		} else if kind == seqDNA || kind == seqRNA {
			allowed := "ACGTN"
			if kind == seqRNA {
				allowed = "ACGUN"
			}
			if !strings.ContainsRune(allowed, r) {
				invalid = append(invalid, fmt.Sprintf("pos %d: %q", i+1, r))
			}
		}
		if len(invalid) >= 5 {
			break
		}
	}

	var b strings.Builder
	fmt.Fprintf(&b, "Sequence analysis (in silico)\n")
	fmt.Fprintf(&b, "Type: %s\n", kindStr)
	fmt.Fprintf(&b, "Length: %d residues/bases\n", len(seq))
	if truncated {
		fmt.Fprintf(&b, "Note: truncated to %d characters for analysis\n", maxLen)
	}
	if len(invalid) > 0 {
		fmt.Fprintf(&b, "Invalid characters (sample): %s\n", strings.Join(invalid, ", "))
	}
	if kind == seqDNA {
		rc := reverseComplementDNA(seq)
		if len(rc) <= 120 {
			fmt.Fprintf(&b, "Reverse complement: %s\n", rc)
		} else {
			fmt.Fprintf(&b, "Reverse complement (first 60): %s...\n", rc[:60])
		}
	}
	if kind == seqProtein && len(seq) <= maxFoldLength() {
		fmt.Fprintf(&b, "Eligible for fold_protein tool (length <= %d aa)\n", maxFoldLength())
	} else if kind == seqProtein {
		fmt.Fprintf(&b, "Too long for fold_protein (max %d aa); analyze or truncate first\n", maxFoldLength())
	}
	fmt.Fprintf(&b, "\nDisclaimer: research/education only; not clinical advice.\n")
	return b.String(), nil
}

File: internal/packs/testdata/official/life-sciences/assets/hub/routes/sequence.py
"""Sequence analysis for biology sidecar."""
from __future__ import annotations

import re
from typing import Any

DNA = set("ACGTN")
RNA = set("ACGUN")
PROTEIN = set("ACDEFGHIKLMNPQRSTVWY")


def _normalize(raw: str) -> str:
    lines = []
    for line in raw.splitlines():
        line = line.strip()
        if not line or line.startswith(">") or line.startswith(";"):
            continue
        lines.append(line)
    return "".join(lines).upper()


def _classify(seq: str) -> str:
    if not seq:
        return "unknown"
    dna = rna = protein = invalid = 0
    for ch in seq:
        if ch in "ACGT":
            dna += 1
        elif ch == "U":
            rna += 1
        elif ch == "N":
            dna += 1
            rna += 1
        elif ch == "*":
            protein += 1
        elif ch in PROTEIN:
            protein += 1
        else:
            invalid += 1
    if invalid > max(1, len(seq) // 10):
        return "unknown"
    if protein and protein >= dna and protein >= rna:
        return "protein"
    if rna > dna and rna > 0:
        return "rna"
    if dna > 0:
        return "dna"
    return "unknown"


def _reverse_complement_dna(seq: str) -> str:
    comp = {"A": "T", "T": "A", "C": "G", "G": "C", "N": "N"}
    return "".join(comp.get(c, c) for c in reversed(seq))


def analyze_sequence(body: dict, settings: dict, pack_dir: str) -> dict[str, Any]:
    raw = str(body.get("sequence", "")).strip()
    seq = _normalize(raw)
    if not seq:
        raise ValueError("no sequence found (paste FASTA or raw sequence)")
    max_len = int(body.get("max_length") or 10000)
    truncated = False
    if len(seq) > max_len:
        seq = seq[:max_len]
        truncated = True
    kind = _classify(seq)
    out: dict[str, Any] = {
        "type": kind,
        "length": len(seq),
        "truncated": truncated,
        "summary": f"Sequence analysis (in silico)\nType: {kind}\nLength: {len(seq)}",
    }
    if kind == "dna":
        rc = _reverse_complement_dna(seq)
        out["reverse_complement"] = rc[:120] + ("..." if len(rc) > 120 else "")
    if kind == "protein":
        max_fold = int(body.get("max_fold_length") or 400)
        out["fold_eligible"] = len(seq) <= max_fold
    out["disclaimer"] = "research/education only; not clinical advice"
    return out
